docs/RFCS/schema_gossip.md

```diff
@@ -0,0 +1,34 @@
+- Feature Name: schema_distribution
+- Start Date: 2015-07-20
+- RFC PR:
+- Cockroach Issue:
+
+# Summary
+
+This RFC suggests implementing eventually-consistent replication of the SQL schema to all Cockroach nodes in a cluster using gossip.
+
+# Motivation
+
+In order to support performant SQL queries, each gateway node must be able to address the data requested by the query. Today this requires the node to read `TableDescriptor`s from the KV map, which we believe (though we haven't measured) will cause a substantial performance hit which we can eliminate by actively gossiping the necessary metadata.
+
+# Detailed design
+
+Distribution of the schema metadata will be modeled after the current configuration distribution mechanism. Whenever a new `TableDescriptor` is written to the KV map, the node holding the leader lease for that key's range will add the `TableDescriptor` to gossip under a key derived from the `TableDescriptor`'s `Name` with a TTL TBD.
+
+On the read side, `sqlserver.Server.getTableDesc`'s behaviour will change such that it reads from gossip rather than the KV store. This will require plumbing a referece to the `Gossip` instance down to the `sqlserver.Server` instance.
+
+# Drawbacks
+
+This is slightly more complicated than the current implementation. Because the schema is eventually-consistent, how do we know when migrations are done? We'll have to count on the TTL, which feels a little dirty.
+
+# Alternatives
+
+We could stick to the current implementation, perhaps using inconsistent reads for better performance. This will be strictly less performant than the gossip approach but will be more optimal in that nodes will only need to carry in memory schema information that they themselves need.
+
+We could have a special metadata range which all nodes have a replica of. This would probably result in unacceptable read and write times which induce lots of network traffic.
+
+We could implement this on top of non-voting replicas (which we don't yet support). That would give us eventual consistency without having to outside of raft, but enforcement of schema freshness remains an open question.
```

Comment:
This whole process seems premature.  We don't know how much of a performance hit we're getting from this, right from the motivation section: `which we believe (though we haven't measured) will cause a substantial performance hit`

The alternative I would suggest would be to store the table descriptors in each range that contains any data form the table. In order to makes changes to the schema, we can do a SI transaction (so it's not blocked by reads) and this will also ensure that we don't ever make two changes at the same time.

This has several advantages:
- consistent (no need for sorting out the TTL)
- puts the schema next to the data
- doesn't increase network traffic
- uses systems we already have built
- can allow us to keep a history of schema version if needed
- if needed, we can build a specialty cash in the nodes for them

Disadvantages:
- slightly more complex to implement - but I don't think it's significantly worse than the other alternatives
- schema changes could require a fairly large transaction, which could fail - schema changes in very large tables require locking and other tricks normally, so I don't see this as a huge drawback and I've had them fail on me numerous times
- only nodes with the data have the schema - but I don't see how that's a major disadvantage
- small chance of anomalies (different versions of the schema on different ranges) - but we should be able to detect them easily enough)